### PR TITLE
PACA-656: Use DSL to generate all RIF entities (Part 1/2)

### PR DESCRIPTION
**Pull request by brianburton**
**JIRA Ticket:**
[PACA-656](https://jira.cms.gov/browse/PACA-656)

**User Story or Bug Summary:**

As a BFD engineer I want to have a single code generator for both RIF and RDA entities and transformer/parsers.  As a first step develop DSL mapping files for all RIF entities and use the DSL maven plugin to generate entity classes instead of the annotation processor.  Continue to generate RIF parsers using annotation processor.

---

### What Does This PR Do?

This PR includes the following changes:

- Add mapping DSL files for all RIF entities that are currently generated by the annotation processor.
- Modify annotation processor to disable generation of entity and column class files.
- Configure the `pom.xml` file in the `bfd-model-rif` project to use the DSL plugin to generate entities and sample DDL.
- Minor fix to DSL entity generation plugin to fix adding an `EqualsHashCode.Include` to fields in classes that do not have an `EqualsHashCode` at the class level.
- Minor fix to set the default `fetchType` on `OneToMany` joins to `LAZY` instead of `EAGER`.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.
* Generated entities must be compatible with those generated by the annotation processor.
* All existing unit and integration tests must pass with no code changes when using the DSL generated entities.
* Hibernate schema compatibility checks must pass with the new entities.

The generated entities can have small changes as long as they do not break compatibility.

This PR focuses only on the Hibernate entities themselves.  A follow on PR will deal with adopting the DSL generated parsers.  While the DSL in this PR contains transformation specs those do not have to be complete yet.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by KyRobbins — APPROVED** [review]
Nice, another nice step to see (outside of the scope of this PR) would be to use the field constants from Lombok for the JPA specs rather than generated meta models.

File: apps/bfd-model/bfd-model-rif/mappings/BeneficiaryHistory.yaml [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+      name: mbiEffectiveDate
+      sqlType: date
+    - comment: MBI Obsolete Date
+      dbName: efctv_end_dt
+      name: mbiObsoleteDate
+      sqlType: date
+    - comment: HICN_UNHASHED (https://bluebutton.cms.gov/resources/variables/hicnUnhashed)
+      dbName: hicn_unhashed
+      name: hicnUnhashed
+      sqlType: varchar(11)
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-model/bfd-model-rif/mappings/BeneficiaryHistory.yaml:130]: Good catch.  This is a change from the old entities.  The reason for this one is that the column in the database is actually a `varchar(11)` so I made the model correspond to the table.

**Review by daryll — DISMISSED** [review]
looks good to me so far

**Review by KyRobbins — APPROVED** [review]

**Review by daryll — APPROVED** [review]
lgtm!

thx u @brianburton @mackec1

File: apps/bfd-model/bfd-model-rif/mappings/BeneficiaryMonthly.yaml [hunk 2]
```diff
@@ -0,0 +1,88 @@
+mappings:
```
Comment (mackec1) [inline apps/bfd-model/bfd-model-rif/mappings/BeneficiaryMonthly.yaml:1]: +++++++
Gets rid of the code that used to do this....so much better!!!

**Review by mackec1 — APPROVED** [review]
Awesome work...this will make much easier, especially allowing diffs to be performed on the yaml instead of the binary xslx. Thx for doing this.

**Review by mackec1 — APPROVED** [review]
BRAVO....this is so good. Thx for doing this.

**Review by cbrunefearless — APPROVED** [review]
LGTM

**Review by lsmitchell — APPROVED** [review]
Very glad to work towards removing the csv file for these mappings; had a few implementation questions but nothing blocking, just curiosity.

(Didn't fine-tooth comb through the yamls, since it was discussed that these have been picked through and tested already.)

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/GenerateEntitiesFromDslMojo.java [hunk 3]
```diff
@@ -415,7 +415,9 @@ FieldDefinition createFieldDefinitionForJoin(RootBean root, MappingBean mapping,
         FieldSpec.builder(fieldType, join.getFieldName()).addModifiers(Modifier.PRIVATE);
     if (mapping.getTable().isPrimaryKey(join)) {
       fieldSpec.addAnnotation(Id.class);
-      fieldSpec.addAnnotation(EqualsAndHashCode.Include.class);
+      if (mapping.getTable().isEqualsNeeded()) {
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/GenerateEntitiesFromDslMojo.java:418]: `equalsNeeded` is a property in the`TableBean` set by whoever created the DSL file.  So ultimately the answer to that question is whoever created the file.

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/model/ColumnBean.java [hunk 4]
```diff
@@ -25,23 +24,23 @@
...
    * length.
    */
   private static final Pattern CharacterLengthRegex =
-      Pattern.compile("char\\((\\d+)\\)", Pattern.CASE_INSENSITIVE);
+      Pattern.compile("(var)?char(\\((\\d+|max)\\))?", Pattern.CASE_INSENSITIVE);
   /**
    * Group number in {@link ColumnBean#CharacterLengthRegex} that contains the numeric length value.
    */
-  private static final int CharacterLengthGroup = 1;
+  private static final int CharacterLengthGroup = 3;
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/model/ColumnBean.java:43]: That's just how groups work in Java (well, all programming languages I've used) work.  Group 0 is the entire matched string, numbers 1+ are specific match groups.

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/model/ColumnBean.java [hunk 5]
```diff
@@ -131,17 +130,19 @@ public int computeMinLength(int defaultValue) {
...
-    } else if (sqlType.equalsIgnoreCase("varchar(max)")) {
-      return Integer.MAX_VALUE;
-    } else {
-      return 0;
+    var length = 0;
+    final var matcher = CharacterLengthRegex.matcher(Strings.nullToEmpty(sqlType));
+    if (matcher.matches()) {
+      final var lengthString = matcher.group(CharacterLengthGroup);
+      if (Strings.isNullOrEmpty(lengthString)) {
+        length = 1;
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/model/ColumnBean.java:138]: `varchar` in SQL is a string of length 0 or 1.  `varchar(max)` is a clob.  `varchar(n)` is a string of length n.  It's just SQL.

**Pull request merged**